### PR TITLE
Add rate limiting to Kafka ingester

### LINKED ISSUE #1100: Add a rate limiter to jaeger-ingester
## Requirement - what kind of business use case are you trying to solve?
Limit the rate of writes to the storage backend when using jaeger-ingester. 

When there is a failure of the storage backed, the ingester can consume and write to the backing store at rates higher than the backing store can handle, resulting in a failure in storage nodes,  leading to more chaos. We've seen this happen with Cassandra, and is probably an issue with any storage engine that doesn't gracefully degrade. 

## Problem - what in Jaeger blocks you from solving the requirement?
none

## Proposal - what do you suggest to solve the problem or improve the existing situation?
Adding a configurable rate limiter to the Kafka reader will enable operators to set a safe throughput target. 

## Any open questions to address
What would be a sensible default value?

### PR DESCRIPTION
**Pull request by isaachier**
- Resolves #1100.

- Add rate-limiting channel to Kafka ingester.
- Fix typo in consumer.

### CODE REVIEW

File: cmd/ingester/app/consumer/consumer.go [hunk 1]
```diff
@@ -52,12 +56,18 @@ type consumerState struct {
 
 // New is a constructor for a Consumer
 func New(params Params) (*Consumer, error) {
+	var rateLimiter *time.Ticker
```
Comment (vprithvi) [inline cmd/ingester/app/consumer/consumer.go]: Could you encapsulate the rate limiter in it's own struct?

File: cmd/ingester/app/consumer/consumer_test.go [hunk 2]
```diff
@@ -86,9 +86,17 @@ func newConsumer(
 	factory metrics.Factory,
 	topic string,
 	processor processor.SpanProcessor,
-	consumer consumer.Consumer) *Consumer {
+	consumer consumer.Consumer,
+	maxReadsPerSecond float64) *Consumer {
 
 	logger, _ := zap.NewDevelopment()
+
+	var rateLimiter *time.Ticker
```
Comment (vprithvi) [inline cmd/ingester/app/consumer/consumer_test.go]: Could you add more test cases? 
For e.g., 
If the rate limiter is configured to 0, does it allow any requests?
If the rate limiter is configured to 10qps, does it allow 10 requests?
If the rate limiter is configured to 1qps, and it receives 10 requests, does it allow the 10 requests in 10 seconds?

Comment (isaachier) [reply cmd/ingester/app/consumer/consumer_test.go]: Ya definitely. I agree this deserves direct testing.

File: cmd/ingester/app/consumer/consumer.go (hunk 1 above)
Comment (isaachier) [reply cmd/ingester/app/consumer/consumer.go]: Yes

**Comment by isaachier** [comment]
Something is up with the ES test case.

File: cmd/ingester/app/flags.go [hunk 3]
```diff
@@ -54,13 +56,16 @@ const (
 	DefaultParallelism = 1000
 	// DefaultEncoding is the default span encoding
 	DefaultEncoding = EncodingProto
+	// DefaultMaxReadsPerSecond is the default max reads per second for the span processor
+	DefaultMaxReadsPerSecond = 0.0
```
Comment (black-adder) [inline cmd/ingester/app/flags.go]: semantically this is wrong; if you pass in 0, you'd expect the consumer to do no work but this value does the exact opposite. Create a new const `DisableRateLimit = 0.0` in rate_limiter.go and use that here instead.

Comment (yurishkuro) [reply cmd/ingester/app/flags.go]: you can also consider value 0 to be "undefined" and thus not causing any rate limit. Certainly from the CLI API point of view, you should not need to pass any flags to run without a rate limit, and if you do want a rate limit you pass a non-0 value.

File: cmd/ingester/app/flags.go [hunk 4]
```diff
@@ -85,6 +90,10 @@ func AddFlags(flagSet *flag.FlagSet) {
 		ConfigPrefix+SuffixEncoding,
 		DefaultEncoding,
 		fmt.Sprintf(`The encoding of spans ("%s" or "%s") consumed from kafka`, EncodingProto, EncodingJSON))
+	flagSet.String(
```
Comment (yurishkuro) [inline cmd/ingester/app/flags.go]: why is it a string and not a number?

File: cmd/ingester/app/consumer/consumer.go [hunk 5]
```diff
@@ -100,6 +105,7 @@ func (c *Consumer) handleMessages(pc sc.PartitionConsumer) {
 	var msgProcessor processor.SpanProcessor
 
 	for msg := range pc.Messages() {
+		c.rateLimiter.await()
```
Comment (yurishkuro) [inline cmd/ingester/app/consumer/consumer.go]: Alternatively you could wrap `pc.Messages()` channel into another channel that limits the rate of produced messages.  Then your reading code wouldn't be exposed to the rate limiting API, and you can bypass wrapping the channel if the rate parameter is 0.

File: cmd/ingester/app/flags.go (hunk 3 above)
Comment (isaachier) [reply cmd/ingester/app/flags.go]: I was using `-1` for this, but used `0` instead seeing as Go default initialization makes this hard to get wrong. I do not mind using another value if you'd like. Not sure what makes the most sense here.

File: cmd/ingester/app/flags.go (hunk 4 above)
Comment (isaachier) [reply cmd/ingester/app/flags.go]: That's a great question but I based this snippet on the parallelism flag above (also strangely a string). I can use `Float64` instead.

File: cmd/ingester/app/consumer/consumer.go (hunk 5 above)
Comment (isaachier) [reply cmd/ingester/app/consumer/consumer.go]: So I thought about that. I thought it might be a performance issue to add another channel. Better idea would be to benchmark both. Will get back to you.

File: cmd/ingester/app/consumer/rate_limiter.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package consumer
+
+import (
+	"time"
+)
+
+// Rate limiter using ticker explained here: https://gobyexample.com/rate-limiting.
```
[the remark this answers is not in the document]
Comment (isaachier) [reply cmd/ingester/app/consumer/rate_limiter.go]: Ya I wanted to give credit but I guess I should add a summary in case that happens.

File: cmd/ingester/app/consumer/rate_limiter.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+func newRateLimiter(creditsPerSecond float64) *rateLimiter {
+	if creditsPerSecond <= 0 {
+		return &rateLimiter{}
+	}
+	interval := time.Nanosecond * time.Duration(float64(time.Second.Nanoseconds())/creditsPerSecond)
+	return &rateLimiter{ticker: time.NewTicker(interval)}
+}
+
+func (r *rateLimiter) await() {
```
[the remark this answers is not in the document]
Comment (isaachier) [reply cmd/ingester/app/consumer/rate_limiter.go]: Right. I checked Guava's rate limiter to find a name and saw `acquire` (https://github.com/google/guava/blob/master/guava/src/com/google/common/util/concurrent/RateLimiter.java). My own issue was that `acquire` implies `release`, but I guess I'll follow Guava here.

File: cmd/ingester/app/consumer/rate_limiter_test.go [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+	var m sync.Mutex
+	counter := 0
+	for i := 0; i < iterations; i++ {
+		rateLimiter.await()
+		m.Lock()
+		counter++
+		m.Unlock()
+	}
+	require.Equal(t, counter, iterations)
```
[the remark this answers is not in the document]
Comment (isaachier) [reply cmd/ingester/app/consumer/rate_limiter_test.go]: The issue I actually saw was that the rate limiting **did not work** on each iteration, but only as a totality. For example, I had tested that each tick was at least 10 ms apart, and had outright assertion failures because the actual tick came at ~9.64 ms, etc. My guess is the clock is not the same, or the ticker guarantees are much more lax than I had thought.

File: cmd/ingester/app/consumer/consumer.go [hunk 9]
```diff
@@ -99,7 +103,18 @@ func (c *Consumer) handleMessages(pc sc.PartitionConsumer) {
 	msgMetrics := c.newMsgMetrics(pc.Partition())
 	var msgProcessor processor.SpanProcessor
 
-	for msg := range pc.Messages() {
+	// "Cast" channel from chan *sarama.ConsumerMessage to chan interface{}.
```
Comment (vprithvi) [inline cmd/ingester/app/consumer/consumer.go]: Could you elaborate which problem we are solving here? 
The casting seems to happen on L117, and L107 is creating a new channel to copy all messages.

File: cmd/ingester/app/consumer/consumer.go [hunk 10]
```diff
@@ -99,7 +103,18 @@ func (c *Consumer) handleMessages(pc sc.PartitionConsumer) {
...
+	msgChannel := make(chan interface{})
+	go func() {
+		defer close(msgChannel)
+		for msg := range pc.Messages() {
+			msgChannel <- msg
+		}
+	}()
+
+	rateLimiter := newRateLimiter(msgChannel, c.maxReadsPerSecond)
+	for v := range rateLimiter.C {
```
Comment (vprithvi) [inline cmd/ingester/app/consumer/consumer.go]: In theory, wouldn't this perform worse than the previous variant which called `ratelimiter.await` in this loop?

Comment (isaachier) [reply cmd/ingester/app/consumer/consumer.go]: In theory, yes. This is somewhat hard to determine, but I can attempt a benchmark of old vs. new version in the case where no rate limiting is applied (compares await with immediate return against wrapping inner channel). Otherwise, we will just be benchmarking the rate limit, which is useless.

File: cmd/ingester/app/consumer/consumer.go (hunk 9 above)
Comment (isaachier) [reply cmd/ingester/app/consumer/consumer.go]: Definitely will add more detail to comments. A channel such as `chan *sarama.ConsumerMessage` cannot be cast to `chan interface{}`, so I need to forward from the source to the destination channel.

File: cmd/ingester/app/consumer/rate_limiter.go [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+package consumer
+
+import (
+	"sync"
+	"time"
+)
+
+// rateLimiter wraps a channel to provide rate limiting.
+type rateLimiter struct {
```
Comment (vprithvi) [inline cmd/ingester/app/consumer/rate_limiter.go]: IMHO, this rate limiter seems pretty complex in that it requires a bunch of channels, a wait group, and a Ticker.

Seeing that this rate limiter is only being used from a single threaded context, and that we are most interested in the stable behavior of the rate limiter,  what advantages does it have over an simpler implementation like one described in https://stackoverflow.com/questions/667508/whats-a-good-rate-limiting-algorithm ?

**Comment by isaachier** [comment]
As I began this channel wrapping, I realized it would probably be worse than a simple rate limiter. I'm glad I went through with it, if only to develop my own understand of channels. Nonetheless, I think I'll revert this latest change in favor of the simpler API.

File: cmd/ingester/app/consumer/rate_limiter_test.go [hunk 12]
```diff
@@ -57,12 +57,12 @@ func TestRateLimiterNoLimit(t *testing.T) {
 	r := newRateLimiter(creditsPerSecond)
 	defer r.Stop()
 
-	const maxInterval = time.Microsecond
+	const maxInterval = time.Millisecond
 	acquireStart := time.Now()
 	for i := 0; i < iterations; i++ {
```
Comment (yurishkuro) [inline cmd/ingester/app/consumer/rate_limiter_test.go]: I don't see any references to `r` in this loop - what is it testing?

Comment (isaachier) [reply cmd/ingester/app/consumer/rate_limiter_test.go]: Good catch. Guess I missed the most important line here.

File: cmd/ingester/app/consumer/rate_limiter.go [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+	done        chan struct{}
+}
+
+// newRateLimiter constructs a rate limiter with the provided creditsPerSecond.
+// If creditsPerSecond is less than zero, no rate limiting will be
+// provided, and all calls to this object will return immediately. If
+// creditsPerSecond is exactly zero, calls will block indefinitely or until Stop
+// is called.
+func newRateLimiter(creditsPerSecond float64) *rateLimiter {
+	var r *rateLimiter
```
Comment (black-adder) [inline cmd/ingester/app/consumer/rate_limiter.go]: I think we can simplify this a bit. Initialize empty tickChannel and ticker at the top and go through the elseif craziness. At the bottom, initialize the rateLimiter. Should reduce the amount of `done:        make(chan struct{})` everywhere. Also, I prefer switch statement here instead of if else (just personal preference).

File: cmd/ingester/app/consumer/rate_limiter.go [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+	if creditsPerSecond < 0 {
```
Comment (black-adder) [inline cmd/ingester/app/consumer/rate_limiter.go]: define -1 and 0 as consts somewhere. Ala
```
disableRateLimiting = -1
noProgress = 0 <- awful name, think of something better
```

File: cmd/ingester/app/builder/builder.go [hunk 15]
```diff
@@ -70,10 +70,11 @@ func CreateConsumer(logger *zap.Logger, metricsFactory metrics.Factory, spanWrit
...
 	consumerParams := consumer.Params{
-		InternalConsumer: saramaConsumer,
-		ProcessorFactory: *processorFactory,
-		Factory:          metricsFactory,
-		Logger:           logger,
+		InternalConsumer:  saramaConsumer,
+		ProcessorFactory:  *processorFactory,
+		Factory:           metricsFactory,
+		Logger:            logger,
+		MaxReadsPerSecond: options.MaxReadsPerSecond,
```
Comment (black-adder) [inline cmd/ingester/app/builder/builder.go]: ah, I think I've made a mistake when scoping this. This static config means we'd have to redeploy the ingester/indexer everytime we want to throttle vs using a dynamic config value. I think we have a way to plug in dynamic config somewhere else in this repo, let's copy that here.

File: cmd/ingester/app/consumer/rate_limiter.go (hunk 13 above)
Comment (isaachier) [reply cmd/ingester/app/consumer/rate_limiter.go]: Sure. I forgot Go allowed switches with conditions instead of values.

File: cmd/ingester/app/builder/builder.go (hunk 15 above)
Comment (isaachier) [reply cmd/ingester/app/builder/builder.go]: Yes. We discussed this, yet I focused on the rate limiter and hadn't thought about refactoring the config. This might be harder to do than it sounds... I have to check other places we did this.

File: cmd/ingester/app/consumer/rate_limiter_test.go (hunk 12 above)
Comment (isaachier) [reply cmd/ingester/app/consumer/rate_limiter_test.go]: Fixed.

File: cmd/ingester/app/consumer/rate_limiter.go (hunk 13 above)
Comment (vprithvi) [reply cmd/ingester/app/consumer/rate_limiter.go]: I still feel that this general approach is more complex than using a simple token bucket without involving channels. 
What benefits do we gain by doing this? 

See https://github.com/uber-go/ratelimit/blob/c15da02342779cb6dc027fc95ee2277787698f36/ratelimit.go for an example - IMHO it is more readable.

Comment (yurishkuro) [reply cmd/ingester/app/consumer/rate_limiter.go]: I tend to agree. That rate limiter just sleeps until the next token is available. Is that worse than using channels for signaling?

File: cmd/ingester/app/consumer/rate_limiter_test.go [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+	acquireStart := time.Now()
+	for i := 0; i < iterations; i++ {
+		acquireEnd := time.Now()
+		interval := acquireEnd.Sub(acquireStart)
+		require.Truef(t, interval <= maxInterval, "Expected no significant delay before first receive, actual delay = %v, max expected delay = %v", interval, maxInterval)
+		acquireStart = time.Now()
+	}
+}
+
+func TestRateLimiterInfiniteLimit(t *testing.T) {
```
Comment (vprithvi) [inline cmd/ingester/app/consumer/rate_limiter_test.go]: Could you also add a benchmark comparing a empty loop to a loop with a ratelimiter with creditsPerSec set to `math.MaxFloat64`.  I'm curious about the overhead introduced and wanted to verify that it runs > 10k qps.

File: cmd/ingester/app/consumer/rate_limiter.go (hunk 13 above)
Comment (isaachier) [reply cmd/ingester/app/consumer/rate_limiter.go]: I will admit the rate limiter here is also less accurate. The ticker may drop ticks if we aren't receiving from the tick channel frequently enough. My issue with redoing the token bucket is that we should ideally just use the one in the client, and potentially migrate it to jaeger-lib. This is the second time I have reimplemented a token bucket in the project (throttling was first).

Comment (isaachier) [reply cmd/ingester/app/consumer/rate_limiter.go]: There is a slightly different semantic, so I guess "less accurate" is not a good way of putting it. The token bucket allows for bursts if enough credits have accumulated. A ticker will not allow for any bursts.

Comment (isaachier) [reply cmd/ingester/app/consumer/rate_limiter.go]: Oh man... I already did that in https://github.com/uber/jaeger-lib/pull/35.

Comment (vprithvi) [reply cmd/ingester/app/consumer/rate_limiter.go]: This is excellent - could we reuse that ratelimiter?

Comment (isaachier) [reply cmd/ingester/app/consumer/rate_limiter.go]: If we weren't expecting a blocking rate limiter, I would have just reused that. The blocking involves calculating a interval and waiting for a timer. The existing rate limiter does not expose its credit balance, making this more difficult.

File: cmd/ingester/app/consumer/rate_limiter.go (hunk 11 above)
Comment (isaachier) [reply cmd/ingester/app/consumer/rate_limiter.go]: Done.

File: cmd/ingester/app/consumer/rate_limiter.go (hunk 13 above)
Comment (vprithvi) [reply cmd/ingester/app/consumer/rate_limiter.go]: Is it very difficult to extend the ratelimiter in `jaeger-lib` to do this?

**Comment by isaachier** [comment]
Hey @vprithvi I have rebased onto latest master and I see the new deadlock detector uses a ticker to calculate throughput, ensuring no consumer is stuck. However, this causes the consumer unit test to fail when rate limiting is applied with the deadlock detector running. Rate limiting inherently changes the throughput. So should I add something to the deadlock detector to allow for rate limiting?

**Comment by yurishkuro** [comment]
why should they interfere?  rate limiting is meant to slow things down, but not stop them completely, while deadlock detector detects complete lack of throughput.

**Comment by yurishkuro** [comment]
Yes, it is exactly the throughput monitor, but it's time horizon should be orders of magnitude larger than the throughput limits of the rate limiter. I.e. the former - no messages in 1min, the latter - up to 1000 msg/sec. So I don't see a fundamental conflict here. You may have clashes in the tests if tests bring those time limits much closer to each other, but they can still be adjusted to be different orders of magnitude.

**Comment by isaachier** [comment]
Right. I'll look at the test case that panics and see what I can do.

**Comment by isaachier** [comment]
Never mind, I was being stupid. The rate limiter change made negative `maxReadsPerSecond` become undefined behavior. I think it was blocking while filling a negative balance. Fixing now.

**Comment by isaachier** [comment]
Any ideas why TestElasticsearchStorage consistently fails in Travis?

File: cmd/ingester/app/consumer/consumer.go [hunk 17]
```diff
@@ -110,8 +114,16 @@ func (c *Consumer) handleMessages(pc sc.PartitionConsumer) {
 
 	deadlockDetector := c.deadlockDetector.startMonitoringForPartition(pc.Partition())
 	defer deadlockDetector.close()
-
+	var rateLimiter *rateLimiter
+	if c.maxReadsPerSecond > 0 {
```
Comment (vprithvi) [inline cmd/ingester/app/consumer/consumer.go]: Why allow the `rateLimiter` to be `nil`?
If we consolidate the logic for when `maxReadsPerSecond <= 0` into the rateLImiter, then we can remove the `if` in L124

File: cmd/ingester/app/consumer/rate_limiter.go [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+// limitations under the License.
+
+package consumer
+
+import (
+	"math"
+	"time"
+)
+
+const cost = 1.0
```
Comment (vprithvi) [inline cmd/ingester/app/consumer/rate_limiter.go]: What is the purpose of this?

File: cmd/ingester/app/consumer/rate_limiter.go [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+
+// Stop stops any ongoing Acquire operation, which exits immediately.
+func (r *rateLimiter) Stop() {
+	close(r.event)
+}
+
+// Update replaces the current rate limiter configuration in a thread-safe
+// manner. This function will panic if Stop has already been called on the
+// rateLimiter.
+func (r *rateLimiter) Update(creditsPerSecond, maxBalance float64) {
```
Comment (vprithvi) [inline cmd/ingester/app/consumer/rate_limiter.go]: Where is this used?

File: cmd/ingester/app/consumer/rate_limiter.go [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+// Acquire causes rateLimiter to spend a credit and return immediately,
+// or block until a credit becomes available. Returns true if credits were
+// acquired. Returns false if Stop was called before credits were acquired,
+// interrupting the Acquire call.
+// Note: Acquire has no corresponding release operation.
+// N.B. This rate limiter implementation is not thread-safe. Do not call Acquire
+// from more than one goroutine.
+func (r *rateLimiter) Acquire() bool {
+	for r.updateBalance() < cost {
+		timer := time.NewTimer(r.calculateWait())
```
Comment (vprithvi) [inline cmd/ingester/app/consumer/rate_limiter.go]: What is the benefit of having `Acquire` being interruptible? 
Creating a new timer whenever the balance is less than the cost is quite convoluted - an alternative is to sleep for the calculated wait time.

File: cmd/ingester/app/consumer/rate_limiter.go (hunk 18 above)
Comment (isaachier) [reply cmd/ingester/app/consumer/rate_limiter.go]: Avoiding the use of a magic number throughout. Honestly, I believe 1 is pretty understandable, but we do something very similar in the Go client so I decided to follow that precedent.

File: cmd/ingester/app/consumer/rate_limiter.go (hunk 19 above)
Comment (isaachier) [reply cmd/ingester/app/consumer/rate_limiter.go]: We are considering (at least internally) the idea of dynamically changing this value if/when we need it.

File: cmd/ingester/app/consumer/consumer.go (hunk 17 above)
Comment (vprithvi) [reply cmd/ingester/app/consumer/consumer.go]: I see - IMHO a cleaner way of solving this would be to use a no-op rateLimiter implementation that is picked when the user doesn't require rate limiting

File: cmd/ingester/app/consumer/rate_limiter.go (hunk 19 above)
Comment (vprithvi) [reply cmd/ingester/app/consumer/rate_limiter.go]: I think we should add it when we have a concrete use case - why does `Update` need to be thread safe when `Acquire` isn't?

File: cmd/ingester/app/consumer/rate_limiter.go (hunk 20 above)
Comment (isaachier) [reply cmd/ingester/app/consumer/rate_limiter.go]: Or I could always use `time.AfterFunc`. I'm not sure it makes sense to implement a blocking rate limiter without the ability to cancel/interrupt any blocking calls. Sounds like it might lead to unexpected hangs in an application down the line. Also, this allows us to handle update channels etc. while we sleep.

File: cmd/ingester/app/consumer/consumer.go (hunk 17 above)
Comment (isaachier) [reply cmd/ingester/app/consumer/consumer.go]: Good idea!

File: cmd/ingester/app/consumer/rate_limiter.go (hunk 19 above)
Comment (isaachier) [reply cmd/ingester/app/consumer/rate_limiter.go]: Thread-safe might be a little exaggerated. I'll correct it to be `May be called concurrently with Acquire`.

In broad terms, I'm trying to use Go's concurrency primitives instead of the sync package. The main thinking here is that Go's condition variables have no `WaitUntil` method, so I cannot easily implement an interruptible wait without falling back on a timer. At that point, I might as well use the timer channel instead of the locking.

File: cmd/ingester/app/consumer/rate_limiter.go (hunk 18 above)
Comment (vprithvi) [reply cmd/ingester/app/consumer/rate_limiter.go]: My bad - I misread

File: cmd/agent/app/httpserver/collector_proxy_test.go [hunk 21]
```diff
@@ -22,7 +22,7 @@ import (
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"github.com/uber/jaeger-lib/metrics"
-	mTestutils "github.com/uber/jaeger-lib/metrics/testutils"
+	"github.com/uber/jaeger-lib/metrics/metricstest"
```
Comment (yurishkuro) [inline cmd/agent/app/httpserver/collector_proxy_test.go]: can we do this in a separate PR?

Comment (yurishkuro) [reply cmd/agent/app/httpserver/collector_proxy_test.go]: also, why do we need to do this at all? The new jaeger-lib is not released as a version afaik.

Comment (isaachier) [reply cmd/agent/app/httpserver/collector_proxy_test.go]: Fair enough. I opened issue #1140 with PR #1141. It would make sense to do this eventually. Meanwhile, I had originally assumed I was using the rate limiter from jaeger-lib, which prompted me to upgrade the library here.

**Comment by isaachier** [comment]
Redoing this in writer instead.
